### PR TITLE
repart: Add support for drop-in overrides

### PR DESCRIPTION
**Pull request by rphibel**

### CODE REVIEW

File: src/partition/repart.c [hunk 1]
```diff
@@ -1407,6 +1420,11 @@ static int context_read_definitions(
                 return log_error_errno(r, "Failed to enumerate *.conf files: %m");
 
         STRV_FOREACH(f, files) {
+                const char *drop_in_dir = strjoina(*f, ".d");
+                _cleanup_strv_free_ char **drop_in_files = NULL;
+                r = conf_files_list_strv(&drop_in_files, ".conf", NULL, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) STRV_MAKE(drop_in_dir));
+                if (r < 0)
+                        return log_error_errno(r, "Failed to enumerate drop-in files: %m");
```
Comment (daandemeyer) [inline src/partition/repart.c]: We generally don't mix declarations with statements, so first, we declare the variables, and then, we follow with the actual logic, so this stuff should move below all the variable declarations.

File: src/partition/repart.c [hunk 2]
```diff
@@ -1407,6 +1420,11 @@ static int context_read_definitions(
                 return log_error_errno(r, "Failed to enumerate *.conf files: %m");
 
         STRV_FOREACH(f, files) {
+                const char *drop_in_dir = strjoina(*f, ".d");
```
Comment (daandemeyer) [inline src/partition/repart.c]: If you read the paragraphs on drop-in units in https://www.freedesktop.org/software/systemd/man/systemd.unit.html, you'll find that drop-ins are not simply searched in the same directory that the main config file is located in, but in all directories that config files are searched in.

So if systemd-repart is looking for configuration files in /usr/lib/repart.d/, etc/repart.d/, and /run/repart.d/, and it finds a config file /etc/repart.d/root-partition.conf, we should then look for drop-in files in /usr/lib/repart.d/root-partition.conf.d/, /etc/repart.d/root-partition.conf.d and /run/repart.d/root-partition.conf.d/.

In other words, for each file, you need to construct a strv of possible dropin directories by iterating over all possible config directories and then pass that list to `conf_files_list_strv()` to get the actual drop-in files.

File: src/partition/repart.c [hunk 3]
```diff
@@ -1417,7 +1435,9 @@ static int context_read_definitions(
                 if (!p->definition_path)
                         return log_oom();
 
-                r = partition_read_definition(p, *f);
+                p->drop_in_files = strv_copy(drop_in_files);
```
Comment (daandemeyer) [inline src/partition/repart.c]: You can replace `strv_copy()` with `TAKE_PTR()`, this will assign `drop_in_files` to `p->drop_in_files` and then set `drop_in_files` to `NULL`. This makes sure that the strv is not freed by `_cleanup_strv_free_`.

File: src/partition/repart.c [hunk 4]
```diff
@@ -1346,6 +1348,17 @@ static int partition_read_definition(Partition *p, const char *path) {
         if (r < 0)
                 return r;
 
+        STRV_FOREACH(drop_in_file, drop_in_files) {
+                r = config_parse(NULL, *drop_in_file, NULL,
```
Comment (poettering) [inline src/partition/repart.c]: nah, no need to hack this up manually, just use `config_parse_many_nulstr`

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
config_parse_many_nulstr() will deal with the drop-on ordering logic for you. Just use that.

File: src/partition/repart.c (hunk 4 above)
Comment (daandemeyer) [reply src/partition/repart.c]: That wouldn't work here right, since we don't use a fixed config file. Shouldn't it be `config_parse_many()` instead?

Comment (poettering) [reply src/partition/repart.c]: sure

File: src/shared/conf-parser.c [hunk 5]
```diff
@@ -602,7 +603,11 @@ int config_parse_many(
         if (r < 0)
                 return r;
 
-        return config_parse_many_files(conf_files, files, sections, lookup, table, flags, userdata, ret_stats_by_path);
+        r = config_parse_many_files(conf_files, files, sections, lookup, table, flags, userdata, ret_stats_by_path);
```
Comment (daandemeyer) [inline src/shared/conf-parser.c:606]: We only set return values on function success, so we should add an error check and return early here if `config_parse_many_files()` failed.

File: src/shared/conf-parser.c [hunk 6]
```diff
@@ -602,7 +603,11 @@ int config_parse_many(
...
+
+        if (ret_dropin_files) *ret_dropin_files = TAKE_PTR(files);
```
Comment (daandemeyer) [inline src/shared/conf-parser.c]: We never put the conditional and statement on the same line, so please add a newline here after the conditional.

File: src/shared/conf-parser.c [hunk 7]
```diff
@@ -602,7 +603,11 @@ int config_parse_many(
...
+
+        return r;
```
Comment (daandemeyer) [inline src/shared/conf-parser.c]: This can become `return 0` after the above comments are addressed.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
The `config_parse_many()` change commit should also be ordered before the repart commit

File: src/partition/repart.c [hunk 8]
```diff
@@ -1336,13 +1338,19 @@ static int partition_read_definition(Partition *p, const char *path) {
...
         };
         int r;
-
-        r = config_parse(NULL, path, NULL,
-                         "Partition\0",
-                         config_item_table_lookup, table,
-                         CONFIG_PARSE_WARN,
-                         p,
-                         NULL);
+        const char* filename = basename(path);
```
Comment (poettering) [inline src/partition/repart.c]: please do not use basename() in new code anymore. Use path_extract_filename()

File: src/partition/repart.c [hunk 9]
```diff
@@ -1407,8 +1415,16 @@ static int context_read_definitions(
...
         STRV_FOREACH(f, files) {
+                char **conf_file_dirs;
+                _cleanup_strv_free_ char **drop_in_files = NULL;
                 _cleanup_(partition_freep) Partition *p = NULL;
 
+                if (directory)
+                        conf_file_dirs = STRV_MAKE(directory);
+                else {
+                        conf_file_dirs = CONF_PATHS_STRV("repart.d");
+                }
```
Comment (daandemeyer) [inline src/partition/repart.c]: We should reduce duplication by only calculating the list of directories to use once, since it's already done for the first call to `conf_files_list_strv()` earlier in this function as well. e.g. I'd do something like the following at the start of this function:

```
const char *const *dirs;

if (directory)
        dirs = STRV_MAKE_CONST(directory);
else
        dirs = (const char* const*) CONF_PATHS_STRV("repart.d");
```

Also remember that single line if statements should not have accompanying braces.

File: src/partition/repart.c [hunk 10]
```diff
@@ -1310,7 +1312,7 @@ static int config_parse_gpt_flags(
         return 0;
 }
 
-static int partition_read_definition(Partition *p, const char *path) {
+static int partition_read_definition(Partition *p, const char *path, char **conf_file_dirs) {
```
Comment (daandemeyer) [inline src/partition/repart.c]: Type of `conf_file_dirs` here should already be `const char *const *`

File: src/partition/repart.c [hunk 11]
```diff
@@ -1416,7 +1416,6 @@ static int context_read_definitions(
 
         STRV_FOREACH(f, files) {
                 char **conf_file_dirs;
-                _cleanup_strv_free_ char **drop_in_files = NULL;
```
Comment (daandemeyer) [inline src/partition/repart.c]: This should be merged into the second commit

File: src/partition/test-repart.sh [hunk 12]
```diff
@@ -249,7 +249,7 @@ EOF
 
 rm -f abc
 
-systemd-repart --definitions="$D/definitions-overrides" --dry-run=yes --empty=create --size=100M --json=pretty abc >"$D/dropin-overrides"
```
Comment (daandemeyer) [inline src/partition/test-repart.sh]: This should be merged into the 3rd commit

File: src/partition/test-repart.sh [hunk 13]
```diff
@@ -222,3 +222,49 @@ echo "### Testing json output ###"
...
+Label=label1
+SizeMaxBytes=32M
+EOF
+
+cat >"$D/definitions-overrides/root.conf.d/override2.conf" <<EOF
+[Partition]
+Label=label2
+EOF
+
+rm -f abc
```
Comment (daandemeyer) [inline src/partition/test-repart.sh]: I'd use `test-drop-in-image` as the name, that's a bit more descriptive than abc

File: src/partition/test-repart.sh [hunk 14]
```diff
@@ -222,3 +222,49 @@ echo "### Testing json output ###"
...
+
+systemd-repart --definitions="$D/definitions-overrides" --dry-run=yes --empty=create --size=100M --json=pretty abc >"$D/dropin-overrides"
```
Comment (daandemeyer) [inline src/partition/test-repart.sh]: You can just assign the output to a bash variable as follows instead of using a file

```
JSON_OUTPUT=$(systemd-repart --definitions="$D/definitions-overrides" --dry-run=yes --empty=create --size=100M --json=pretty abc)
```

This should only capture stdout and not stderr in the variable.

File: src/partition/test-repart.sh [hunk 15]
```diff
@@ -222,3 +222,49 @@ echo "### Testing json output ###"
...
+
+cmp "$D/dropin-overrides" - <<EOF
```
Comment (daandemeyer) [inline src/partition/test-repart.sh]: You can use `diff` instead of `cmp`, it provides better output when things don't match.

To pass strings to `diff`, you can do the following:

```
diff <(echo "$JSON_OUTPUT") - <<EOF
<json>
EOF
```

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks good, a few more comments

File: src/partition/repart.c [hunk 16]
```diff
@@ -1396,13 +1404,16 @@ static int context_read_definitions(
...
         assert(context);
 
         if (directory)
-                r = conf_files_list_strv(&files, ".conf", NULL, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) STRV_MAKE(directory));
+                dirs = STRV_MAKE_CONST(directory);
         else
-                r = conf_files_list_strv(&files, ".conf", root, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) CONF_PATHS_STRV("repart.d"));
+                dirs = (const char* const*) CONF_PATHS_STRV("repart.d");
+
+        r = conf_files_list_strv(&files, ".conf", NULL, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, dirs);
```
Comment (yuwata) [inline src/partition/repart.c]: `root` is dropped mistakenly(?).

**Review by DaanDeMeyer — APPROVED** [review]

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/partition/repart.c (hunk 8 above)
Comment (daandemeyer) [inline src/partition/repart.c]: I'd drop this variable since it's only used once

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
One more nit on top of @yuwata's comment about `root` being mistakenly dropped

File: src/partition/repart.c [hunk 17]
```diff
@@ -1396,13 +1403,18 @@ static int context_read_definitions(
...
-        if (directory)
-                r = conf_files_list_strv(&files, ".conf", NULL, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) STRV_MAKE(directory));
-        else
-                r = conf_files_list_strv(&files, ".conf", root, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) CONF_PATHS_STRV("repart.d"));
+        if (directory) {
+                dirs = STRV_MAKE_CONST(directory);
+                r = conf_files_list_strv(&files, ".conf", NULL, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, dirs);
+        } else {
+                dirs = (const char* const*) CONF_PATHS_STRV("repart.d");
+                r = conf_files_list_strv(&files, ".conf", root, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, dirs);
```
Comment (daandemeyer) [inline src/partition/repart.c]: You can just do this outside of the `if` with a ternary operator to avoid duplication:

```
r = conf_files_list_strv(&files, ".conf", directory ? NULL : root, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, dirs);
```

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: src/partition/repart.c (hunk 16 above)
Comment (rphibel) [reply src/partition/repart.c]: Yes, it was a mistake. It should be fixed now.

**Review by DaanDeMeyer — APPROVED** [review]

**Review by yuwata — APPROVED** [review]

**Comment by yuwata** [comment]
```
stdout:
### Testing systemd-repart --empty=create ###
stderr:
+ set -o pipefail
+ [[ -e /dev/loop-control ]]
+ repart=/builddir/build/BUILD/systemd-251/redhat-linux-build/systemd-repart
+ udevadm=/builddir/build/BUILD/systemd-251/redhat-linux-build/udevadm
+ test -x /builddir/build/BUILD/systemd-251/redhat-linux-build/systemd-repart
+ test -x /builddir/build/BUILD/systemd-251/redhat-linux-build/udevadm
+ PATH=/builddir/.local/bin:/builddir/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin:/sbin:/usr/sbin
++ mktemp --tmpdir --directory test-repart.XXXXXXXXXX
+ D=/tmp/test-repart.51F9OxG74d
+ trap 'rm -rf '\''/tmp/test-repart.51F9OxG74d'\''' EXIT INT QUIT PIPE
+ mkdir -p /tmp/test-repart.51F9OxG74d/definitions
+ SEED=e2a40bf9-73f1-4278-9160-49c031e7aef8
+ echo '### Testing systemd-repart --empty=create ###'
+ /builddir/build/BUILD/systemd-251/redhat-linux-build/systemd-repart /tmp/test-repart.51F9OxG74d/zzz --empty=create --size=1G --seed=e2a40bf9-73f1-4278-9160-49c031e7aef8 --no-pager
/builddir/build/BUILD/systemd-251/src/partition/test-repart.sh: line 25: 60847 Segmentation fault      (core dumped) "$repart" "$D/zzz" --empty=create --size=1G --seed="$SEED" --no-pager
+ rm -rf /tmp/test-repart.51F9OxG74d
```

File: src/partition/repart.c [hunk 18]
```diff
@@ -1396,13 +1403,16 @@ static int context_read_definitions(
...
+        const char *const *dirs;
 
         assert(context);
 
         if (directory)
-                r = conf_files_list_strv(&files, ".conf", NULL, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) STRV_MAKE(directory));
+                dirs = STRV_MAKE_CONST(directory);
         else
-                r = conf_files_list_strv(&files, ".conf", root, CONF_FILES_REGULAR|CONF_FILES_FILTER_MASKED, (const char**) CONF_PATHS_STRV("repart.d"));
+                dirs = (const char* const*) CONF_PATHS_STRV("repart.d");
```
Comment (yuwata) [inline src/partition/repart.c]: You cannot use `dirs` outside of the scope, like `alloca()`.

I think it is not necessary to optimize here. How about to simply use `strv_new()` and `strv_copy()`?

Comment (rphibel) [reply src/partition/repart.c]: Thanks for spotting the issue! I will try your suggestion.

Comment (daandemeyer) [reply src/partition/repart.c]: What about using a ternary instead of an if block? That would also fix the issue.

Comment (rphibel) [reply src/partition/repart.c]: Good idea. Thanks for the suggestion.

Comment (rphibel) [reply src/partition/repart.c]: Good idea. Thanks for the suggestion.

File: src/partition/repart.c [hunk 19]
```diff
@@ -1336,13 +1338,18 @@ static int partition_read_definition(Partition *p, const char *path) {
...
         };
         int r;
-
-        r = config_parse(NULL, path, NULL,
-                         "Partition\0",
-                         config_item_table_lookup, table,
-                         CONFIG_PARSE_WARN,
-                         p,
-                         NULL);
+        const char* dropin_dirname = strjoina(basename(path), ".d");
```
Comment (poettering) [inline src/partition/repart.c]: please do not use basename() in new code anymore. Use path_extract_filename()

File: src/partition/repart.c [hunk 20]
```diff
@@ -2058,7 +2067,8 @@ static int context_dump_partitions(Context *context, const char *node) {
                                 TABLE_UINT64, p->current_padding == UINT64_MAX ? 0 : p->current_padding,
                                 TABLE_UINT64, p->new_padding,
                                 TABLE_STRING, padding_change, TABLE_SET_COLOR, !p->partitions_next && sum_padding > 0 ? ansi_underline() : NULL,
-                                TABLE_STRING, activity ?: "unchanged");
+                                TABLE_STRING, activity ?: "unchanged",
+                                TABLE_STRV, p->drop_in_files);
```
Comment (poettering) [inline src/partition/repart.c:2081]: hmm, what's the point of displaying this? I mean, it's great we read it, but outside of debugging, I doubt this information should show up. Maybe in json output, but certainly not in the regular text output? (if you want to keep this in the json output, then hide it otherwise via table_hide_column_from_display())

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/partition/repart.c (hunk 20 above)
Comment (daandemeyer) [reply src/partition/repart.c:2081]: Does it really hurt to include it in the text output? I figure it's useful for users to be able to see all dropins that apply to some partition config

Comment (poettering) [reply src/partition/repart.c:2081]: at the very least hide the column of not a single partition has drop-ins. That should suffice for me

File: src/partition/repart.c [hunk 21]
```diff
@@ -1336,13 +1338,25 @@ static int partition_read_definition(Partition *p, const char *path) {
...
 
-        r = config_parse(NULL, path, NULL,
-                         "Partition\0",
-                         config_item_table_lookup, table,
-                         CONFIG_PARSE_WARN,
-                         p,
-                         NULL);
+        r = path_extract_filename(path, &filename);
+        if (r < 0)
+                return r;
```
Comment (poettering) [inline src/partition/repart.c]: please error log here.

File: src/partition/repart.c [hunk 22]
```diff
@@ -2058,9 +2077,12 @@ static int context_dump_partitions(Context *context, const char *node) {
                                 TABLE_UINT64, p->current_padding == UINT64_MAX ? 0 : p->current_padding,
                                 TABLE_UINT64, p->new_padding,
                                 TABLE_STRING, padding_change, TABLE_SET_COLOR, !p->partitions_next && sum_padding > 0 ? ansi_underline() : NULL,
-                                TABLE_STRING, activity ?: "unchanged");
+                                TABLE_STRING, activity ?: "unchanged",
+                                TABLE_STRV, p->drop_in_files);
                 if (r < 0)
                         return table_log_add_error(r);
+
+                no_dropin_files &= strv_isempty(p->drop_in_files);
```
Comment (poettering) [inline src/partition/repart.c]: a bitwise `&` on a bool is weird. please use `n_d_f = n_d_f && …`

sooner or later a static analyzer will otherwise complain

**Comment by poettering** [comment]
lgtm

**Comment by DaanDeMeyer** [comment]
CI failure is unrelated
